Zend/zend_operators.c

```diff
@@ -3218,7 +3221,9 @@ ZEND_API zend_string* ZEND_FASTCALL zend_long_to_str(zend_long num) /* {{{ */
 	} else {
 		char buf[MAX_LENGTH_OF_LONG + 1];
 		char *res = zend_print_long_to_buf(buf + sizeof(buf) - 1, num);
-		return zend_string_init(res, buf + sizeof(buf) - 1 - res, 0);
+		zend_string *str =  zend_string_init(res, buf + sizeof(buf) - 1 - res, 0);
+		GC_ADD_FLAGS(str, IS_STR_VALID_UTF8);
```

Comment:
Is this assumption valid even for files with arbitrary charsets/encodings, and arbitrary settings of `zend.multibyte`? Genuinely asking, since I've never used anything but UTF-8.

Reply:
I'm not sure if I totally understand the question.

AFAIK, numbers in UTF-16 would be encoded over two bytes. But looking at the implementation of ``zend_print_long_to_buf()`` it seems to just create an ASCII string and doesn't care about zend_multibyte.

However, my understanding of what zend_multibyte does is to convert the current script file *into* an internal encoding to process the file, which is most likely going to be UTF-8. But that's a good question I don't know the answer. But it seems like we assume anyway that the encoding is going to be ASCII compatible

Reply:
Yeah, it looks like these functions always return ASCII strings.

I don't think this is related at all to files with arbitrary charsets; yes, string literals in such files might use some other text encoding, but this is not about string literals. There is no reason why PHP would use `zend_long_to_str` to create string objects for string literals embedded in a source file.